resources/definitions/fdmprinter.def.json

```diff
@@ -932,13 +946,63 @@
                     "limit_to_extruder": "wall_0_extruder_nr",
                     "settable_per_mesh": true
                 },
+                "topmost_skin_extruder_nr":
+                {
+                    "label": "Roofing Extruder",
+                    "description": "The extruder train used for printing the top most skin. This is used in multi-extrusion.",
+                    "type": "optional_extruder",
+                    "default_value": "-1",
+                    "settable_per_mesh": true,
+                    "settable_per_extruder": false,
+                    "settable_per_meshgroup": true,
+                    "settable_globally": true,
+                    "enabled": "machine_extruder_count > 1 and topmost_skin_layer_count > 0 and top_layers > 0"
+                },
+                "topmost_skin_layer_count":
+                {
+                    "label": "Roofing Layers",
+                    "description": "The number of top most skin layers. Usually only one top most layer is sufficient to generate higher quality top surfaces.",
+                    "default_value": 0,
+                    "minimum_value": "0",
+                    "maximum_value_warning": "top_layers - 1",
+                    "type": "int",
+                    "value": "0 if infill_sparse_density == 100 else 1",
+                    "limit_to_extruder": "topmost_skin_extruder_nr",
+                    "settable_per_mesh": true,
+                    "enabled": "top_layers > 0"
+                },
+                "topmost_skin_pattern":
+                {
+                    "label": "Roofing Pattern",
+                    "description": "The pattern of the top most layers.",
+                    "type": "enum",
+                    "options":
+                    {
+                        "lines": "Lines",
+                        "concentric": "Concentric",
+                        "zigzag": "Zig Zag"
+                    },
+                    "default_value": "lines",
+                    "limit_to_extruder": "topmost_skin_extruder_nr",
+                    "settable_per_mesh": true,
+                    "enabled": "topmost_skin_layer_count > 0 and top_layers > 0"
+                },
+                "topmost_skin_angles":
+                {
+                    "label": "Roofing Line Directions",
+                    "description": "A list of integer line directions to use when the roofing layers use the lines or zig zag pattern. Elements from the list are used sequentially as the layers progress and when the end of the list is reached, it starts at the beginning again. The list items are separated by commas and the whole list is contained in square brackets. Default is an empty list which means use the traditional default angles (45 and 135 degrees).",
+                    "type": "[int]",
+                    "default_value": "[ ]",
```

Comment:
Perhaps it's a better idea to simply inherit from the skin angles, so we won't need to make any exception case for the empty list.